### ISSUE TITLE
nix-build -A win32.hello fails

### ISSUE THREAD

**Issue by pmiddend** [DavidEGrayson/nixcrpkgs#17]
I checked out the repository and ran `nix-build -A win32.hello`. This starts to compile pkgconf and fails at that:

```
*** Downloading ‘https://cache.nixos.org/nar/1vhplaikanxzpfm8asw2ircyg93s52402ay83hxwzns4011z3pgy.nar.xz’ (signed by ‘cache.nixos.org-1’) to ‘/nix/store/41rkfilvfwhjznh2c4jzp2pjh4zyq39s-gcc-6.3.0.tar.bz2’...
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
  0     0    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0checking for gcc... gcc
checking whether the C compiler works... exporting git://git.code.sf.net/p/mingw-w64/mingw-w64 (rev 6de0055f99ed447ec63c1a650a3830f266a808bd) into /nix/store/fv5kvsga3ab278042a47vkn857l77j52-mingw-w64-6de0055
Initialized empty Git repository in /nix/store/fv5kvsga3ab278042a47vkn857l77j52-mingw-w64-6de0055/.git/
yes
checking for C compiler default output file name... a.out
checking for suffix of executables... 
checking whether we are cross compiling... no
checking for suffix of object files... o
checking whether we are using the GNU C compiler... yes
checking whether gcc accepts -g... yes
checking for gcc option to accept ISO C89... none needed
checking whether gcc understands -c and -o together... yes
checking for strlcpy... no
checking for strlcat... no
checking for strndup... yes
checking for strtok_r... yes
checking for a BSD-compatible install... /nix/store/qd55j183ym04y43aam889xkimq704rdx-coreutils-8.29/bin/install -c
checking whether build environment is sane... yes
checking for a thread-safe mkdir -p... /nix/store/qd55j183ym04y43aam889xkimq704rdx-coreutils-8.29/bin/mkdir -p
checking for gawk... gawk
checking whether make sets $(MAKE)... yes
checking for style of include used by make... GNU
checking whether make supports nested variables... yes
  4 95.1M    4 4072k    0     0  4072k      0  0:00:23 --:--:--  0:00:23 7246kgcc3
checking whether make supports nested variables... (cached) yes
checking build system type... x86_64-unknown-linux-gnu
checking host system type... x86_64-unknown-linux-gnu
checking how to print strings... printf
checking for a sed that does not truncate output... /nix/store/zgr9zbw7p8gihn9h0nj8ggkcplrd9g7f-gnused-4.4/bin/sed
checking for grep that handles long lines and -e... /nix/store/rhbm7javg5n5zmhhld7qbi23n5qc0srf-gnugrep-3.1/bin/grep
checking for egrep... /nix/store/rhbm7javg5n5zmhhld7qbi23n5qc0srf-gnugrep-3.1/bin/grep -E
checking for fgrep... /nix/store/rhbm7javg5n5zmhhld7qbi23n5qc0srf-gnugrep-3.1/bin/grep -F
checking for ld used by gcc... /nix/store/6s3dl13ingr21z93dy6zxcc652wdlcrh-gcc-wrapper-6.4.0/bin/ld
checking if the linker (/nix/store/6s3dl13ingr21z93dy6zxcc652wdlcrh-gcc-wrapper-6.4.0/bin/ld) is GNU ld... yes
checking for BSD- or MS-compatible name lister (nm)... no
checking for dumpbin... no
checking for link... link -dump
checking the name lister (nm) interface... BSD nm
checking whether ln -s works... yes
checking the maximum length of command line arguments... 32768
checking how to convert x86_64-unknown-linux-gnu file names to x86_64-unknown-linux-gnu format... func_convert_file_noop
checking how to convert x86_64-unknown-linux-gnu file names to toolchain format... func_convert_file_noop
checking for /nix/store/6s3dl13ingr21z93dy6zxcc652wdlcrh-gcc-wrapper-6.4.0/bin/ld option to reload object files... -r
checking for objdump... no
checking how to recognize dependent libraries... pass_all
checking for dlltool... no
checking how to associate runtime and link libraries... printf %s\n
checking for ar... no
checking for archiver @FILE support... no
checking for strip... no
checking for ranlib... no
checking command to parse nm output from gcc object... failed
checking for sysroot... no
checking for a working dd... /nix/store/qd55j183ym04y43aam889xkimq704rdx-coreutils-8.29/bin/dd
checking how to truncate binary pipes... /nix/store/qd55j183ym04y43aam889xkimq704rdx-coreutils-8.29/bin/dd bs=4096 count=1
../pkgconf-1.0.1/configure: line 6406: /usr/bin/file: No such file or directory
checking for mt... no
checking if : is a manifest tool... no
checking how to run the C preprocessor... gcc -E
checking for ANSI C header files... yes
checking for sys/types.h... yes
checking for sys/stat.h... yes
checking for stdlib.h... yes
checking for string.h... yes
checking for memory.h... yes
checking for strings.h... yes
checking for inttypes.h... yes
checking for stdint.h... yes
checking for unistd.h... yes
checking for dlfcn.h... yes
checking for objdir... .libs
checking if gcc supports -fno-rtti -fno-exceptions... no
checking for gcc option to produce PIC... -fPIC -DPIC
checking if gcc PIC flag -fPIC -DPIC works... yes
checking if gcc static flag -static works... no
checking if gcc supports -c -o file.o... yes
checking if gcc supports -c -o file.o... (cached) yes
checking whether the gcc linker (/nix/store/6s3dl13ingr21z93dy6zxcc652wdlcrh-gcc-wrapper-6.4.0/bin/ld) supports shared libraries... yes
checking whether -lc should be explicitly linked in... no
checking dynamic linker characteristics... GNU/Linux ld.so
checking how to hardcode library paths into programs... immediate
checking whether stripping libraries is possible... no
checking if libtool supports shared libraries... yes
checking whether to build shared libraries... yes
checking whether to build static libraries... yes
checking for special C compiler options needed for large files... no
checking for _FILE_OFFSET_BITS value needed for large files... no
checking whether to enable strict mode... (cached) no
checking how to run the C preprocessor... gcc -E
 14 95.1M   14 13.9M    0     0  13.9M      0  0:00:06  0:00:01  0:00:05 9137kchecking for gcc... (cached) gcc
checking whether we are using the GNU C compiler... (cached) yes
checking whether gcc accepts -g... (cached) yes
checking for gcc option to accept ISO C89... (cached) none needed
checking whether gcc understands -c and -o together... (cached) yes
checking whether ln -s works... yes
checking that generated files are newer than configure... done
configure: creating ./config.status
config.status: creating Makefile
config.status: creating tests/run.sh
config.status: creating libpkgconf/config.h
config.status: executing depfiles commands
config.status: executing libtool commands
  CC       libpkgconf/audit.lo
  CC       libpkgconf/cache.lo
  CC       libpkgconf/pkg.lo
 25 95.1M   25 24.4M    0     0  12.2M      0  0:00:07  0:00:02  0:00:05 9774k  CC       libpkgconf/bsdstubs.lo
  CC       libpkgconf/fragment.lo
  CC       libpkgconf/argvsplit.lo
  CC       libpkgconf/fileio.lo
  CC       libpkgconf/tuple.lo
 37 95.1M   37 35.6M    0     0  11.8M      0  0:00:08  0:00:03  0:00:05 10.0M  CC       libpkgconf/dependency.lo
  CC       libpkgconf/queue.lo
  CCLD     libpkgconf.la
make: *** [Makefile:556: libpkgconf.la] Error 1
builder for ‘/nix/store/n4sn9j9cmshpn67g87jhrb60rzr05dyc-pkgconf-1.0.1.drv’ failed with exit code 2
*** Downloading ‘https://cache.nixos.org/nar/1vhplaikanxzpfm8asw2ircyg93s52402ay83hxwzns4011z3pgy.nar.xz’ (signed by ‘cache.nixos.org-1’) to ‘/nix/store/41rkfilvfwhjznh2c4jzp2pjh4zyq39s-gcc-6.3.0.tar.bz2’...
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
  0     0    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0checking for gcc... gcc
checking whether the C compiler works... exporting git://git.code.sf.net/p/mingw-w64/mingw-w64 (rev 6de0055f99ed447ec63c1a650a3830f266a808bd) into /nix/store/fv5kvsga3ab278042a47vkn857l77j52-mingw-w64-6de0055
Initialized empty Git repository in /nix/store/fv5kvsga3ab278042a47vkn857l77j52-mingw-w64-6de0055/.git/
yes
checking for C compiler default output file name... a.out
checking for suffix of executables... 
checking whether we are cross compiling... no
checking for suffix of object files... o
checking whether we are using the GNU C compiler... yes
checking whether gcc accepts -g... yes
checking for gcc option to accept ISO C89... none needed
checking whether gcc understands -c and -o together... yes
checking for strlcpy... no
checking for strlcat... no
checking for strndup... yes
checking for strtok_r... yes
checking for a BSD-compatible install... /nix/store/qd55j183ym04y43aam889xkimq704rdx-coreutils-8.29/bin/install -c
checking whether build environment is sane... yes
checking for a thread-safe mkdir -p... /nix/store/qd55j183ym04y43aam889xkimq704rdx-coreutils-8.29/bin/mkdir -p
checking for gawk... gawk
checking whether make sets $(MAKE)... yes
checking for style of include used by make... GNU
checking whether make supports nested variables... yes
  4 95.1M    4 4072k    0     0  4072k      0  0:00:23 --:--:--  0:00:23 7246kgcc3
checking whether make supports nested variables... (cached) yes
checking build system type... x86_64-unknown-linux-gnu
checking host system type... x86_64-unknown-linux-gnu
checking how to print strings... printf
checking for a sed that does not truncate output... /nix/store/zgr9zbw7p8gihn9h0nj8ggkcplrd9g7f-gnused-4.4/bin/sed
checking for grep that handles long lines and -e... /nix/store/rhbm7javg5n5zmhhld7qbi23n5qc0srf-gnugrep-3.1/bin/grep
checking for egrep... /nix/store/rhbm7javg5n5zmhhld7qbi23n5qc0srf-gnugrep-3.1/bin/grep -E
checking for fgrep... /nix/store/rhbm7javg5n5zmhhld7qbi23n5qc0srf-gnugrep-3.1/bin/grep -F
checking for ld used by gcc... /nix/store/6s3dl13ingr21z93dy6zxcc652wdlcrh-gcc-wrapper-6.4.0/bin/ld
checking if the linker (/nix/store/6s3dl13ingr21z93dy6zxcc652wdlcrh-gcc-wrapper-6.4.0/bin/ld) is GNU ld... yes
checking for BSD- or MS-compatible name lister (nm)... no
checking for dumpbin... no
checking for link... link -dump
checking the name lister (nm) interface... BSD nm
checking whether ln -s works... yes
checking the maximum length of command line arguments... 32768
checking how to convert x86_64-unknown-linux-gnu file names to x86_64-unknown-linux-gnu format... func_convert_file_noop
checking how to convert x86_64-unknown-linux-gnu file names to toolchain format... func_convert_file_noop
checking for /nix/store/6s3dl13ingr21z93dy6zxcc652wdlcrh-gcc-wrapper-6.4.0/bin/ld option to reload object files... -r
checking for objdump... no
checking how to recognize dependent libraries... pass_all
checking for dlltool... no
checking how to associate runtime and link libraries... printf %s\n
checking for ar... no
checking for archiver @FILE support... no
checking for strip... no
checking for ranlib... no
checking command to parse nm output from gcc object... failed
checking for sysroot... no
checking for a working dd... /nix/store/qd55j183ym04y43aam889xkimq704rdx-coreutils-8.29/bin/dd
checking how to truncate binary pipes... /nix/store/qd55j183ym04y43aam889xkimq704rdx-coreutils-8.29/bin/dd bs=4096 count=1
../pkgconf-1.0.1/configure: line 6406: /usr/bin/file: No such file or directory
checking for mt... no
checking if : is a manifest tool... no
checking how to run the C preprocessor... gcc -E
checking for ANSI C header files... yes
checking for sys/types.h... yes
checking for sys/stat.h... yes
checking for stdlib.h... yes
checking for string.h... yes
checking for memory.h... yes
checking for strings.h... yes
checking for inttypes.h... yes
checking for stdint.h... yes
checking for unistd.h... yes
checking for dlfcn.h... yes
checking for objdir... .libs
checking if gcc supports -fno-rtti -fno-exceptions... no
checking for gcc option to produce PIC... -fPIC -DPIC
checking if gcc PIC flag -fPIC -DPIC works... yes
checking if gcc static flag -static works... no
checking if gcc supports -c -o file.o... yes
checking if gcc supports -c -o file.o... (cached) yes
checking whether the gcc linker (/nix/store/6s3dl13ingr21z93dy6zxcc652wdlcrh-gcc-wrapper-6.4.0/bin/ld) supports shared libraries... yes
checking whether -lc should be explicitly linked in... no
checking dynamic linker characteristics... GNU/Linux ld.so
checking how to hardcode library paths into programs... immediate
checking whether stripping libraries is possible... no
checking if libtool supports shared libraries... yes
checking whether to build shared libraries... yes
checking whether to build static libraries... yes
checking for special C compiler options needed for large files... no
checking for _FILE_OFFSET_BITS value needed for large files... no
checking whether to enable strict mode... (cached) no
checking how to run the C preprocessor... gcc -E
 14 95.1M   14 13.9M    0     0  13.9M      0  0:00:06  0:00:01  0:00:05 9137kchecking for gcc... (cached) gcc
checking whether we are using the GNU C compiler... (cached) yes
checking whether gcc accepts -g... (cached) yes
checking for gcc option to accept ISO C89... (cached) none needed
checking whether gcc understands -c and -o together... (cached) yes
checking whether ln -s works... yes
checking that generated files are newer than configure... done
configure: creating ./config.status
config.status: creating Makefile
config.status: creating tests/run.sh
config.status: creating libpkgconf/config.h
config.status: executing depfiles commands
config.status: executing libtool commands
  CC       libpkgconf/audit.lo
  CC       libpkgconf/cache.lo
  CC       libpkgconf/pkg.lo
 25 95.1M   25 24.4M    0     0  12.2M      0  0:00:07  0:00:02  0:00:05 9774k  CC       libpkgconf/bsdstubs.lo
  CC       libpkgconf/fragment.lo
  CC       libpkgconf/argvsplit.lo
  CC       libpkgconf/fileio.lo
  CC       libpkgconf/tuple.lo
 37 95.1M   37 35.6M    0     0  11.8M      0  0:00:08  0:00:03  0:00:05 10.0M  CC       libpkgconf/dependency.lo
  CC       libpkgconf/queue.lo
  CCLD     libpkgconf.la
make: *** [Makefile:556: libpkgconf.la] Error 1
builder for ‘/nix/store/n4sn9j9cmshpn67g87jhrb60rzr05dyc-pkgconf-1.0.1.drv’ failed with exit code 2
```

**Comment by DavidEGrayson** [comment]
Thanks for letting me know about this.  I can reproduce your problem.

I ran `cat ~/.nix-defexpr/channels/nixpkgs/.git-revision` to see what version of nixpkgs I am using.  My virtual machine was previously on https://github.com/nixos/nixpkgs/commit/b212125b5418fe181e320dbfefd0efc8e7c2e3d5, and the command `nix-build -A native.pkgconf` worked fine for me.  However, when I ran `nix-channel --update` to update nixpkgs, it updated to https://github.com/nixos/nixpkgs/commit/2f6b88551ba76eaf7ec1f0cf9dec02c3ffb284f7 and then the `pkgconf` build stopped working.

So it seems that nixpkgs has made some change that broke nixcrpkgs.

I cloned the nixpkgs repository to `~/nixpkgs`, and ran `export NIX_PATH=/home/david/nixpkgs` to make `nix-build` use that copy of nixpkgs instead of the one from the channel.  Then I went to `~/nixpkgs` and used the `git bisect` command to switch around between different commits and figure out when that project made the first bad commit.  I tested each commit by running `nix-build -A native.pkgconf` from my nixcrpkgs clone.  Most of these builds failed or succeeded pretty quickly, but when building commit https://github.com/nixos/nixpkgs/commit/99806c5e12bd02926981ebf2d0b779c683b801ee, I noticed that it had to build perl and a lot of other things from source.  It might take a while for me to finish doing this git bisect and solve the underlying problem, but I'll keep you updated.

**Comment by DavidEGrayson** [comment]
As a temporary workaround, you can switch back to https://github.com/NixOS/nixpkgs/commit/b212125b5418fe181e320dbfefd0efc8e7c2e3d5, since that is the commit that my virtual machine was previously using and it's good as far as I know.  Just do something like:

```
cd
git clone https://github.com/NixOS/nixpkgs
cd nixpkgs
git checkout b212125b5418fe181e320dbfefd0efc8e7c2e3d5 
cd ..
export NIX_PATH=nixpkgs=/home/pmiddend/nixpkgs
```

**Comment by pmiddend** [comment]
Wow, thanks for your quick reply. Take your time. I'm investigating cross-compiling with NixOS, and your project seems to be the only "solid" source of information, so thanks for that. :)

**Comment by DavidEGrayson** [comment]
Thanks.  I think nixcrpkgs is pretty good myself, but you should know that @Ericson2314 has been working on doing cross-compilation in nixpkgs itself, and I think he contributed this part of the nixpkgs documentation:

https://nixos.org/nixpkgs/manual/#chap-cross

It didn't seem like a good fit for me because I want everything to be statically linked, while the default in nixpkgs is dynamic linking.  And I wanted the code to be much simpler than what is going on nixpkgs so people can read it linearly and understand what's happening.

**Comment by DavidEGrayson** [comment]
With `git bisect`, I found that the last known good commit of nixpkgs is https://github.com/NixOS/nixpkgs/commit/4f869bccc14fb2fa19df130e76c022765ecda924.  

From commit https://github.com/NixOS/nixpkgs/commit/8e557ed2c58e6ce48a8d05dbc57ef84e98b4cecd (which introduces binutils-wrapper) to https://github.com/NixOS/nixpkgs/commit/99806c5e12bd02926981ebf2d0b779c683b801ee I get a different error that prevents me from even starting the build of pkgconf, usually something like:

```text
cat: /nix/store/j3gqfbda3l23s7h28kplf702d34k99y6-bootstrap-gcc-wrapper/nix-support/libc-ldflags: No such file or directory
cat: /nix/store/j3gqfbda3l23s7h28kplf702d34k99y6-bootstrap-gcc-wrapper/nix-support/libc-ldflags-before: No such file or directory
/nix/store/s1z9vkib4xrxn0rwva9hic4g4bx94mjb-builder.sh: line 61: extraLDFlags[@]: unbound variable
```

Then at commit https://github.com/NixOS/nixpkgs/commit/12e0672d8848852677c6c77a63e2c793ad7a266b, I can reproduce the error you reported.

**Comment by DavidEGrayson** [comment]
I found a concrete example of a breaking change from nixpkgs.  If I build the `binutils` of nixpkgs at commit 4f869bc (the last good commit), the thing that gets built has a `bin` directory with lots of executables:

```
addr2line  dlltool  ld       objcopy  srconv   windres
ar         dllwrap  ld.bfd   objdump  strings
as         dwp      ld.gold  ranlib   strip
c++filt    elfedit  nlmconv  readelf  sysdump
coffdump   gprof    nm       size     windmc
```

But if I build binutils from 12e0672, it's actually just a binutils wrapper and it only has a small fraction of the utilities available in its `bin` directory:

```
as  ld  ld.bfd  ld.gold
```

It seems that with @Ericson2314's new binutils wrapper we can't just count on having the needed utilities in the `bin` directory of `binutils`, we have to do some extra setup steps to add the original binutils to the PATH.  I would really like that PATH to be computed at evalulation time instead of taking care of it at build time with shell scripts, so I will look into how to do that.  Maybe we don't need the binutils wrapper at all, we'll see.

**Comment by DavidEGrayson** [comment]
I fixed this in 98837d0bdd5ee5db184e125f4eee5b5dfe2eabf5 by adding `binutils-unwrapped` to the PATH.  I tested that `nix-build -A win32.hello` works again.
